[DeviceMesh] Add DeviceMesh.manual_seed API

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140450



follows RFC #140301
Also, following a suggestion from @yifuwang, starts off by implementing the API just for DeviceMesh.  This is easier to validate (ensure that dims are 'orthogonal') and we encourage adoption of DeviceMesh anyway for new use cases and especially multi-dimensional parallelism use cases.  We can always follow up with an API in torch.distributed namespace that works for ProcessGroup if we want.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o @tianyu-l @XilunWu

wconstab committed:
[DeviceMesh] Add DeviceMesh.manual_seed API

follows RFC #140301

[ghstack-poisoned]

wz337 commented on torch/distributed/device_mesh.py:
```diff
@@ -924,6 +924,68 @@ def _flatten(self, mesh_dim_name: Optional[str] = None) -> "DeviceMesh":
 
             return _mesh_resources.create_flatten_mesh(self, mesh_dim_name)
+        
+        def manual_seed(self, seed: int, unique_dims: Tuple[Union[str, int], ...]=()):
```
Oh. Could we do start it off as private as we discussed? `_manual_seed`

wz337 commented on torch/distributed/tensor/_random.py:
```diff
@@ -164,6 +164,8 @@ def __init__(self, device_type: str = "cuda"):
         super().__init__(device_type)
         # synchronize RNG state using rank 0's current one
         rng_state = self._device_handle.get_rng_state().to(device_type)
+        print(f"rank {dist.get_rank()} broadcast RNG state")
+        # torch.distributed.breakpoint()
```
nit: remove?

wz337 commented:
Overall it looks good to me! Thanks!

wconstab committed:
Update on "[DeviceMesh] Add DeviceMesh.manual_seed API"


follows RFC #140301

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o tianyu-l XilunWu

[ghstack-poisoned]

wconstab committed:
Update on "[DeviceMesh] Add DeviceMesh.manual_seed API"


follows RFC #140301

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o tianyu-l XilunWu

[ghstack-poisoned]

wconstab committed:
Update on "[DeviceMesh] Add DeviceMesh.manual_seed API"


follows RFC #140301

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o tianyu-l XilunWu

[ghstack-poisoned]

wconstab committed:
Update on "[DeviceMesh] Add DeviceMesh.manual_seed API"




follows RFC #140301
Also, following a suggestion from yifuwang, starts off by implementing the API just for DeviceMesh.  This is easier to validate (ensure that dims are 'orthogonal') and we encourage adoption of DeviceMesh anyway for new use cases and especially multi-dimensional parallelism use cases.  We can always follow up with an API in torch.distributed namespace that works for ProcessGroup if we want.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o tianyu-l XilunWu

[ghstack-poisoned]

XilunWu reviewed (approved):
one caveat: the offset calculation seems a bit problematic. Everything besides looks good to me. I'm thinking of merging this piece of logic into DTensor random or keep it for ProcessGroup.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -925,6 +925,85 @@ def _flatten(self, mesh_dim_name: Optional[str] = None) -> "DeviceMesh":
+                provided, all values must be ints and are interpreted as dim IDs.  No duplicates are allowed.  Dims that
+                are not specified are assumed to be replicated.  An empty 'unique_dims' is valid, e.g for a DDP trainer.
+            """
+            if self != _mesh_resources.get_root_mesh(self):
+                # technically it would be OK to call this API from any submesh that contained all of the unique_dims,
+                # but its probably good practice to always do it from the root mesh?
+                # Another API design could be that if unique_dims is empty, we assume all dims are unique, and users
+                # can just call `fsdp_tp_mesh.manual_seed()` to get the same result as `root_mesh.manual_seed(fsdp,tp)`
+                raise RuntimeError(
+                    "Cannot call manual seed from a submesh, please use the root mesh"
+                )
```
I had the same thoughts.

wz337 replied:
Either way is fine with me. 
For slicing, we are not allowing submesh slicing. 
On the other hand, we have `get_rank()` (call on root mesh) and `get_local_rank()`(call on submesh).

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -925,6 +925,85 @@ def _flatten(self, mesh_dim_name: Optional[str] = None) -> "DeviceMesh":
+            rank_seed = seed
+            prev_size = 1
+            for dim_name in reversed(unique_dims):
+                if isinstance(dim_name, str):
+                    dim = self.__getitem__(dim_name)
+                else:
+                    # dim = TODO
+                    raise NotImplementedError(
+                        "dim-indices are not yet supported for unique_dims"
+                    )
```
hmm, get_local_rank() would return a dim? or a rank?

XilunWu replied:
it returns a process' relative coordinate on a given dim (i.e. a `int`). I believe this is exactly what you need.

wconstab replied:
oh, yes, this is exactly what i need. it will simplify my if/else. let me try.

XilunWu replied:
I see, let's keep the current code then

wconstab committed:
Update on "[DeviceMesh] Add DeviceMesh.manual_seed API"




follows RFC #140301
Also, following a suggestion from yifuwang, starts off by implementing the API just for DeviceMesh.  This is easier to validate (ensure that dims are 'orthogonal') and we encourage adoption of DeviceMesh anyway for new use cases and especially multi-dimensional parallelism use cases.  We can always follow up with an API in torch.distributed namespace that works for ProcessGroup if we want.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o tianyu-l XilunWu

[ghstack-poisoned]

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -925,6 +925,85 @@ def _flatten(self, mesh_dim_name: Optional[str] = None) -> "DeviceMesh":
+                        "dim-indices are not yet supported for unique_dims"
+                    )
+                rank_seed += prev_size * dim.get_local_rank()
+                prev_size = dim.size()
```
fixed this and added a 8gpu test where i could actually test this.

XilunWu replied:
yeah MTPG helps because we don't need real collectives here.

XilunWu commented on test/distributed/test_device_mesh.py:
```diff
@@ -270,6 +270,29 @@ def test_set_mesh_dim_group_options(self):
         # Fake pg only have BackendType as BackendType::CUSTOM.
         self.assertEqual(mesh.get_group(1)._get_backend_name(), "custom")
 
+    @with_comms(eager_init=True)
```
we can use `@with_comms` unless you do need to eagerly init, since the fix (https://github.com/pytorch/pytorch/pull/139637) has been merged.

wconstab replied:
i was getting a lot of nccl warnings about device_id not being specified that went away when i added eager_init=True

wconstab commented:
No longer planning to implement this API after deciding to first harden existing DTensor manual_seed API and make that work with pipeline parallelism directly in torchtitan by calling DTensor manual_seed with different seeds on different PP ranks.
